client/task_runner_test.go

```diff
@@ -195,3 +198,65 @@ func TestTaskRunner_SaveRestoreState(t *testing.T) {
 	}
 }
 */
+
+func TestTaskRunner_HandleDiscovery(t *testing.T) {
+	// Create the task runner
+	_, tr := testTaskRunner()
+
+	// Modify the job so we have predictable results
+	tr.allocID = "1"
+	tr.jobID = "job1"
+	tr.taskGroup = "group1"
+	tr.task.Name = "web"
+
+	// Assign dynamic ports. The ReservedPorts are overloaded for dynamic
+	// ports also, so we set those here as well.
+	tr.task.Resources.Networks[0].DynamicPorts = []string{"http", "https"}
+	tr.task.Resources.Networks[0].ReservedPorts = []int{80, 443}
+
+	// Create the mock discovery layer
+	providers := []discovery.Factory{discovery.NewMockDiscovery}
+	disc, err := discovery.NewDiscoveryLayer(providers, tr.config, tr.logger, tr.config.Node)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	tr.discovery = disc
+	mp := disc.Providers[0].(*discovery.MockDiscovery)
+
+	// Register the task with discovery
+	tr.handleDiscovery(false)
+
+	// Check the result
+	expect := map[string]int{
+		"job1.group1.web:1":       0,
+		"job1.group1.web.http:1":  80,
+		"job1.group1.web.https:1": 443,
```

Comment:
I got your point, but what I am trying to raise here is that tasks spun up with Docker Host mode, won't have an IP/port explicit at Docker inspect, or Nomad metadatas, so how there is no way to automatic discover in some cases, unless explicit configured.

Lets say I have my redis example container spun up with docker host network mode.
The redis server will be listening at 0.0.0.0:6379 on a HostOS that can have a single or multiple IP addresses. In this scenario, would be nice to have a feature to specify at the task the discovery portion where I can tell the IP addresses and ports I want to publish for this task. 

I understand you efforts to make the discovery "transparent" for the user, but since Nomad now supports Docker Host, want to make sure we are covering this mode on service discovery as well, allowing users to custom configure which IP/ports to register for a service(s) when docker network host is used, since in some specific scenarios like HostOS with multiple IPs, and services listening on 0.0.0.0(any), I can't find a way for the code to know better than me which IP I want to register at consul.